### PR TITLE
Add Buffer Latency Metric

### PR DESCRIPTION
**Pull request by kkondaka**
### Description
Add Latency Metric to Buffers
 
### Issues Resolved
Resolves #[Issue number to be closed when this PR is merged]
 
### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
  - [ ] New functionality has javadoc added
- [ X] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by graytaylor0 — DISMISSED** [review]

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/buffer/KafkaBuffer.java [hunk 1]
```diff
@@ -70,7 +70,7 @@ public KafkaBuffer(final PluginSetting pluginSetting, final KafkaBufferConfig ka
                        final AcknowledgementSetManager acknowledgementSetManager,
                        final ByteDecoder byteDecoder, final AwsCredentialsSupplier awsCredentialsSupplier,
                        final CircuitBreaker circuitBreaker) {
-        super(kafkaBufferConfig.getCustomMetricPrefix().orElse(pluginSetting.getName()), pluginSetting.getPipelineName());
+        super(kafkaBufferConfig.getCustomMetricPrefix().orElse(pluginSetting.getName()+"buffer"), pluginSetting.getPipelineName());
```
Comment (dlvenable) [inline data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/buffer/KafkaBuffer.java:73]: What will this name look like?

```
kafkabuffer
```

Is there a better name we can use here?

File: data-prepper-api/src/main/java/org/opensearch/dataprepper/model/buffer/AbstractBuffer.java [hunk 2]
```diff
@@ -142,6 +148,16 @@ public Map.Entry<Collection<T>, CheckpointState> read(int timeoutInMillis) {
         return readResult;
     }
 
+    protected void updateLatency(Collection<T> records) {
+        for (T rec : records) {
+            if (rec instanceof Record) {
+                Record<Event> record = (Record<Event>)rec;
```
Comment (dlvenable) [inline data-prepper-api/src/main/java/org/opensearch/dataprepper/model/buffer/AbstractBuffer.java:154]: Not all buffers hold records at the moment.

We should change these lines to support non-events. Also, I think you don't need to cast to `DefaultEventHandle`.

```
Object data = rec.getData();
if(data instanceof Event) {
  Event event = (Event) data;
  Instant receivedTime = event.getEventHandle().getInternalOriginationTime();
  latencyTimer.record(Duration.between(receivedTime, Instant.now()));
}
```

File: data-prepper-api/src/main/java/org/opensearch/dataprepper/metrics/MetricNames.java [hunk 3]
```diff
@@ -86,4 +86,6 @@ private MetricNames() {}
      * Delimiter used to separate path components in metric names.
      */
     public static final String DELIMITER = ".";
+
+    public static final String LATENCY = "latency";
```
Comment (dlvenable) [inline data-prepper-api/src/main/java/org/opensearch/dataprepper/metrics/MetricNames.java:90]: Should we use a different name here? The term latency is ambiguous as there are numerous places where latency exists.

It appears that this is the time before events are read from the buffer. Some name ideas:

* `bufferReadLatency`
* `readLatency`
* `timeEventIdle`
* `atRestTime` (to put it in opposition to in-flight)

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/buffer/KafkaBuffer.java (hunk 1 above)
Comment (kkondaka) [reply data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/buffer/KafkaBuffer.java:73]: Yeah. Right now the name looks like "kafka". I am changing it to "kafkabuffer"

Comment (dlvenable) [reply data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/buffer/KafkaBuffer.java:73]: I'm wondering if we can have buffer in the base name to start with.

So for example:

```
my-pipeline.kafka.buffer.readLatency
```

Perhaps by changing

```
public static final String READ_LATENCY = "readLatency";
```

to

```
public static final String READ_LATENCY = "buffer.readLatency";
```

Thoughts?

**Review by dlvenable — CHANGES_REQUESTED** [review]

**Review by dlvenable — DISMISSED** [review]

**Review by dlvenable — APPROVED** [review]

**Review by graytaylor0 — APPROVED** [review]

**Pull request merged**